devtools/x/src/lint/guppy.rs

```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
+    ("move-vm-natives", "diem-crypto"),
+    ("move-vm-runtime", "diem-state-view"),
+    ("move-vm-runtime", "diem-logger"),
+    ("move-vm-runtime", "diem-infallible"),
+    ("move-vm-runtime", "diem-crypto"),
+    ("move-vm-types", "diem-crypto"),
+    ("module-generation", "diem-types"),
+    ("test-generation", "language-e2e-tests"),
+    ("test-generation", "diem-vm"),
+    ("test-generation", "diem-types"),
+    ("test-generation", "diem-state-view"),
+    ("test-generation", "diem-logger"),
+    ("test-generation", "diem-config"),
+    ("disassembler", "diem-types"),
+    ("move-cli", "vm-genesis"),
+    ("move-cli", "diem-vm"),
+    ("move-cli", "diem-types"),
+    ("move-cli", "diem-state-view"),
+    ("move-cli", "diem-framework"),
+    ("move-cli", "datatest-stable"),
+    ("move-coverage", "diem-types"),
+    ("resource-viewer", "diem-types"),
+    ("resource-viewer", "diem-state-view"),
+    ("vm", "num-variants"),
+    ("vm", "diem-proptest-helpers"),
+    ("vm", "diem-crypto"),
+];
+
+const DIEM_CRATES_IN_LANGUAGE: &[&str] = &[
+    "diem-framework",
+    "diem-events-fetcher",
+    "diem-validator-interface",
+    "diem-transaction-replay",
+    "diem-writeset-generator",
+    "diem-vm",
+    "language-e2e-testsuite",
+    "ir-testsuite",
+    "language-e2e-tests",
+    "functional-tests",
+    "genesis-viewer",
+    "vm-genesis",
+    "transaction-builder",
+    "transaction-builder-generator",
+];
+
+// Determine if a crate is considered a Move crate or Diem crate.
+//
+// Current criterion:
+//   1. All crates outside language are considered Diem crates.
+//   2. All crates inside language are considered Move crates, unless marked otherwise.
+fn is_move_crate(crate_path: &str, crate_name: &str) -> bool {
+    if crate_path.starts_with("language/") {
+        if DIEM_CRATES_IN_LANGUAGE.contains(&crate_name) {
+            return false;
+        }
+        return true;
+    }
+
+    false
+}
+
+fn is_existing_move_to_diem_dep(name1: &str, name2: &str) -> bool {
+    EXISTING_MOVE_TO_DIEM_DEPS
+        .iter()
+        .any(|(a, b)| *a == name1 && *b == name2)
+}
+
+impl Linter for MoveCratesDontDependOnDiemCrates {
+    fn name(&self) -> &'static str {
+        "move-crates-dont-depend-on-diem-crates"
+    }
+}
+
+impl PackageLinter for MoveCratesDontDependOnDiemCrates {
+    fn run<'l>(
+        &self,
+        ctx: &PackageContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        let metadata = ctx.metadata();
+
+        let crate_name = metadata.name();
+        let crate_path = metadata.source().to_string();
+
+        if is_move_crate(&crate_path, crate_name) {
+            for direct_dep in metadata.direct_links() {
+                let dep = direct_dep.to();
+                let dep_name = dep.name();
+
+                if dep.in_workspace()
+                    && dep_name != "diem-workspace-hack" // Exclude `diem-workspace-hack` to reduce noise.
```

Comment:
Hmm, you could avoid hardcoding it through https://github.com/diem/diem/blob/be4811ee515407175be79cbfcc333fa43ae548d9/devtools/x/src/lint/guppy.rs#L226-L233, but that's honestly much more complicated than it should be. I should fix this.

Reply:
Moved it into x.toml.